src/main/scala/lms/transformation/mpi_nccl_transformation/distributedTensor2MPINCCLKernels.scala

```diff
@@ -0,0 +1,252 @@
+
+      // choose the last two dimensions as dim0 and dim1
+      val dim0_shape = input_shape(input_shape.size - 2)
+      val dim1_shape = input_shape(input_shape.size - 1)
+      val dim0_stride = dim0_shape
+      val dim1_stride = 1
+
+      val input_size = input_shape.fold(1) { _ * _ }
+      val output = gpu_array(input_size, manifest[Float], myNCCLRank)
+
+      val maskedFillKernel = cudaMaskedFill[Float](false)
+      FunOps11(maskedFillKernel).apply(
+        Wrap[Array[Float]](input_tensor), 
+        Wrap[Array[Float]](output.x),
+        Wrap[Array[Int]](mask_tensor),
+        unit[Float](value),
+        unit[Int](dim0_shape),
+        unit[Int](dim1_shape),
+        unit[Int](dim0_stride),
+        unit[Int](dim1_stride),
+        unit[Int](input_size),
+        dim3(unit[Int]((input_size + 511)/512)),
+        dim3(unit[Int](512))
+      )
+
+      output.x
+
+    case Node(s, "tensor_maskedfill_bwd", Backend.Const(tt: TensorType)::Backend.Const(anno:Anno)::(doutput:Backend.Sym)::(mask:Backend.Sym)::_, _) =>
+      implicit val pos = Adapter.oldSourceMap(s)
+
+      // get input info and transform input tensors
+      val doutput_shape = tensor_shape(doutput, useOldMetadata = true)
+      val doutput_tensor = get_operand(doutput, anno)
+      val mask_tensor = get_operand(mask, anno)
+
+      // choose the last two dimensions as dim0 and dim1
+      val dim0_shape = doutput_shape(doutput_shape.size - 2)
+      val dim1_shape = doutput_shape(doutput_shape.size - 1)
+      val dim0_stride = dim0_shape
+      val dim1_stride = 1
+
+      val doutput_size = doutput_shape.fold(1) { _ * _ }
+      val dinput = gpu_array(doutput_size, manifest[Float], myNCCLRank)
+
+      val maskedFillGradKernel = cudaMaskedFillGrad[Float](false)
+      FunOps10(maskedFillGradKernel).apply(
+        Wrap[Array[Float]](doutput_tensor), 
+        Wrap[Array[Float]](dinput.x),
+        Wrap[Array[Int]](mask_tensor),
+        unit[Int](dim0_shape),
+        unit[Int](dim1_shape),
+        unit[Int](dim0_stride),
+        unit[Int](dim1_stride),
+        unit[Int](doutput_size),
+        dim3(unit[Int]((doutput_size + 511)/512)),
+        dim3(unit[Int](512))
+      )
+
+      dinput.x
+    
+    case Node(s, "tensor_logsoftmax", Backend.Const(tt: TensorType)::Backend.Const(anno:Anno)::(input:Backend.Sym)::_, _) =>
+      implicit val pos = Adapter.oldSourceMap(s)
+
+      // get input info and transform input tensors
+      val input_shape = tensor_shape(input, useOldMetadata = true)
+      val input_tensor = get_operand(input, anno)
+
+      val outer_size = input_shape.init.fold(1) { _ * _ }
+      val last_dim_size = input_shape.last
+      val input_size = outer_size * last_dim_size
+
+      val output = gpu_array(input_size, manifest[Float], myNCCLRank)
+
+      val softmaxKernel = cudaSoftmax[Float](true)
+
+      FunOps6(softmaxKernel).apply(
+        Wrap[Array[Float]](input_tensor), 
+        Wrap[Array[Float]](output.x),
```

Comment:
I see that several of the cases are not yet doing `anno match`, but it is OK to merge it as is for now. I will introduce the pass that splits the IRs first.